server/module_hash_by_height.go

```diff
@@ -0,0 +1,125 @@
+Example:
+	appd module-hash-by-height 16841115,
+`,
+		Args: cobra.ExactArgs(1), // Ensure exactly one argument is provided
+		RunE: func(cmd *cobra.Command, args []string) error {
+			heightToRetrieveString := args[0]
+
+			serverCtx := GetServerContextFromCmd(cmd)
+
+			height, err := strconv.ParseInt(heightToRetrieveString, 10, 64)
+			if err != nil {
+				return err
+			}
+
+			commitInfoForHeight, err := getModuleHashesAtHeight(serverCtx, appCreator, height)
+			if err != nil {
+				return err
+			}
+
+			outputFormat, err := cmd.Flags().GetString(flags.FlagOutput)
+			if err != nil {
+				return err
+			}
+
+			switch outputFormat {
+			case flags.OutputFormatJSON:
+				jsonOutput, err := json.MarshalIndent(commitInfoForHeight, "", "  ")
+				if err != nil {
+					return err
+				}
+				cmd.Println(string(jsonOutput))
+			case flags.OutputFormatText:
+				fallthrough
+			default:
+				cmd.Println(commitInfoForHeight.String())
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringP(flags.FlagOutput, "o", flags.OutputFormatText, "Output format (text|json)")
+
+	return cmd
+}
+
+func getModuleHashesAtHeight[T servertypes.Application](svrCtx *Context, appCreator servertypes.AppCreator[T], height int64) (*storetypes.CommitInfo, error) {
+	home := svrCtx.Config.RootDir
+	db, err := openDB(home, GetAppDBBackend(svrCtx.Viper))
+	if err != nil {
+		return nil, fmt.Errorf("error opening DB, make sure osmosisd is not running when calling this query: %w", err)
+	}
+	app := appCreator(svrCtx.Logger, db, nil, svrCtx.Viper)
+	rms, ok := app.CommitMultiStore().(*rootmulti.Store)
+	if !ok {
+		return nil, fmt.Errorf("expected rootmulti.Store, got %T", app.CommitMultiStore())
+	}
+
+	commitInfoForHeight, err := rms.GetCommitInfo(height)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new slice of StoreInfos for storing the modified hashes.
+	storeInfos := make([]storetypes.StoreInfo, len(commitInfoForHeight.StoreInfos))
+
+	for i, storeInfo := range commitInfoForHeight.StoreInfos {
+		// Convert the hash to a hexadecimal string.
+		hash := strings.ToUpper(hex.EncodeToString(storeInfo.CommitId.Hash))
+
+		// Create a new StoreInfo with the modified hash.
+		storeInfos[i] = storetypes.StoreInfo{
+			Name: storeInfo.Name,
+			CommitId: storetypes.CommitID{
+				Version: storeInfo.CommitId.Version,
+				Hash:    []byte(hash),
+			},
+		}
+	}
+
+	// Sort the storeInfos slice based on the module name.
+	sort.Slice(storeInfos, func(i, j int) bool {
+		return storeInfos[i].Name < storeInfos[j].Name
+	})
+
+	// Create a new CommitInfo with the modified StoreInfos.
+	commitInfoForHeight = &storetypes.CommitInfo{
+		Version:    commitInfoForHeight.Version,
+		StoreInfos: storeInfos,
+	}
+
+	return commitInfoForHeight, nil
+}
+
+func openDB(rootDir string, backendType dbm.BackendType) (dbm.DB, error) {
```

Comment:
This can be removed in favour of  [`OpenDB`](https://github.com/cosmos/cosmos-sdk/blob/main/server/util.go#L475`)

Reply:
https://github.com/cosmos/cosmos-sdk/pull/20779/commits/9c01f7c20da57a27bead8ce4689a1d948691d06e